### DISCUSSION TITLE
DNS Turbulent Flow: Follow-up Implementation

### DISCUSSION THREAD

**Discussion by goinea** [mfem/mfem#2576] [General]
Hello @jandrej ,
I have the following paraview results for an edit of the 3d flow over a cylinder file. I replace the mesh with a periodic channel mesh by extending the periodic cube from other mfem files. Do you have any suggestions for getting better channel flow results before going any further
[Turb_Results.zip](https://github.com/mfem/mfem/files/7245963/Turb_Results.zip)
?

**Comment by jandrej** [comment]
Hi,

can you elaborate more on what problem you are trying to solve or a specific result that you want to reproduce?

**Reply by goinea** [reply]
Dr.Julian,
This is a fully developed turbulent channel flow with Re at 590 per the paper: https://www.researchgate.net/publication/243777258_Direct_Numerical_Simulation_of_Turbulent_Channel_Flow_up_to_Re590. The results are to be similar to these.

**Reply by goinea** [reply]
@jandrej

**Reply by jandrej** [reply]
In the branch `navier-updates` you will be able to find an example for the turbulent channel flow here https://github.com/mfem/mfem/blob/navier-updates/miniapps/navier/navier_turbchan.cpp. You can adjust the Reynolds number to your liking but be careful with the initial time steps as well as mesh sizing.

Similar to the references, the mesh is generated with a stretching to accommodate to the boundary layer. The initial flow profile is initialized with the Reichardt function.

**Reply by goinea** [reply]
@jandrej Thank you! I will run it and may ask you a few questions about it later.

**Reply by goinea** [reply]
@jandrej What version of hypre, metis, and mfem did you use? I keep getting a quadrature points verification failure without editing the code? I am using hypre-2.20.0, metis 5, and mfem 4.3.

Error: Verification failed: (Q1D <= MQ) is false:
 Quadrature rules with more than 8 1D points are not supported!
 ... in function: void mfem::internal::quadrature_interpolator::TensorDerivatives(int, int, const mfem::DofToQuad&, const mfem::Vector&, mfem::Vector&) [with mfem::QVectorLayout VL = mfem::QVectorLayout::byNODES]
 ... in file: fem/qinterp/grad_by_nodes.cpp:117

**Reply by jandrej** [reply]
You need to either increase the number of allowed quadrature points in the mentioned file (simply increase the number). Alternatively you can decrease the polynomial order in the example.

I also recommend updating the branch because I changed the initial condition to something more reliable.
<img width="953" alt="Screen Shot 2021-10-04 at 12 10 55 PM" src="https://user-images.githubusercontent.com/5412886/136800424-491410d6-2f64-44c5-972a-c5b23a81c6f1.png">

**Reply by goinea** [reply]
@jandrej Thanks, got it going. Did you run this on a pc or did you remote into a supercomputer? Either way how long was the time?

**Reply by jandrej** [reply]
This was run on an HPC machine on 512 mpi ranks until a simulation time of t=10 or so. It takes a while (~minutes) even on a big machine.

**Reply by Watchmojo** [reply]
I have been waiting on this channel flow using MFEM for a few years. I wish I had time and knowledge to make this. Thank you very much for sharing. @jandrej

**Reply by goinea** [reply]
@jandrej I am running on an HPC with 10 nodes on 400 cores and after about t =6 I get an: 
Assertion failed: (IsFinite(nom)) is false:
 --> nom = inf
 ... in function: virtual void mfem::CGSolver::Mult(const mfem::Vector&, mfem::Vector&) const
 ... in file: /gpfs/u/home/ODLC/ODLCgnsd/barn/dcs/mfem/linalg/solvers.cpp:643

This was the first version of the turb_channel code sent, would using the newest one fix this? Did you get this error?

**Reply by Watchmojo** [reply]
I added the following to compile navier_turbchan
-------------------------------------------------------------
navier_mms-test-par: navier_turbchan
        @$(call mfem-test,$<, $(RUN_MPI), Navier,-cr)
------------------------------------------------------------------
but it returned this error
------------------------------------------------------------
navier_turbchan.cpp(115): error: no instance of constructor "mfem::Vector::Vector" matches the argument list
            argument types are: ({...})
     Vector x_translation({Lx, 0.0, 0.0});
                          ^

navier_turbchan.cpp(116): error: no instance of constructor "mfem::Vector::Vector" matches the argument list
            argument types are: ({...})
     Vector z_translation({0.0, 0.0, Lz});
--------------------------------------------------------------------
Could you please share your makefile to compile navier_turbchan? Thank you.

**Reply by Watchmojo** [reply]
> @jandrej I am running on an HPC with 10 nodes on 400 cores and after about t =6 I get an: Assertion failed: (IsFinite(nom)) is false: --> nom = inf ... in function: virtual void mfem::CGSolver::Mult(const mfem::Vector&, mfem::Vector&) const ... in file: /gpfs/u/home/ODLC/ODLCgnsd/barn/dcs/mfem/linalg/solvers.cpp:643
> 
> This was the first version of the turb_channel code sent, would using the newest one fix this? Did you get this error?

Maybe the DNS run was diverged? reduce time step?

**Reply by jandrej** [reply]
For any build errors please open a separate issue.
